### PR TITLE
Task regression testing

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Task regressions ensure that teachers are giving consistent output from commit to commit. This doesn't ensure they give _correct_ output, just that the output doesn't change.

Task regressions work by creating a particular task with different parameters (fold), and recording the first 5 examples they produce. We also record the number of examples and episodes reported by the teacher. All of this information is stored in a .yml file with the teacher.

On a new commit, we run the task regression to ensure the yml file matches the output generated by the new code. When the yml file changes, and it's intentional, we can run with `pytest --force-regen parlai/tasks/TASK/test.py` to regenerate the yml files. These should then be added to the PR, so that the updated information is included too.

Triggering a breakage of a task regression is *not* a deal breaker for a PR. Rather, is just meant to flag "this changed. Be intentional about that decision."

Works with images, and also does tricks to ensure large numbers of label candidates are not stored.

Remaining steps:
- [x] documentation
- [x] Kill the verify task bit
- [x] Go over with group

**Testing steps**
New CI. Manual testing.

### CODE REVIEW

**Comment by emilydinan** [comment]
holy moly lol

**Comment by stephenroller** [comment]
I know it kills me.

**Comment by stephenroller** [comment]
Hint to whoever eventually reviews this:

image

**Comment by stephenroller** [comment]
Okay, it is done.

File: docs/source/faq.md [hunk 1]
```diff
@@ -2,17 +2,28 @@
...
 
-**Why can't I reproduce the results of an evaluation on a task with a pretrained model?**
+One first thing to check is whether there is a problem with your dictionary or
+token embeddings, because this high perplexity implies that the model is very
+bad at predicting the next token in a string of text.
 
-One common culprit for this is that the flags for that task may not be correctly set. When loading a pretrained checkpoint, all of the parameters for the model itself will be loaded from the model's `.opt` file, but all task-specific parameters will need to be re-specified.
+## I changed my teacher and now its tests won't pass.
 
+Take a careful look at the diff outputs that those tests produce. If the results
```
Comment (EricMichaelSmith) [inline docs/source/faq.md:25]: Hmm should we give some more context about what the regression fixtures are here?

File: docs/source/tutorial_task.md [hunk 2]
```diff
@@ -562,3 +562,56 @@ we could call:
...
+    task = 'myteacher'  # replace with your teacher name
+```
+
+Next, run your test. __It will always fail the first time.__
+
+```bash
+pytest parlai/tasks/myteacher/test.py
+```
+
+You will see many error mesages about "Failed: File not found in data
```
Comment (EricMichaelSmith) [inline docs/source/tutorial_task.md:600]: nit: spelling of "messages"

File: parlai/utils/testing.py [hunk 3]
```diff
@@ -388,91 +389,80 @@ def display_model(opt) -> Tuple[str, str, str]:
...
+            return obj
+        else:
+            return obj
+
+    def _regression(self, data_regression: DataRegressionFixture, datatype: str):
+        import random
+        from parlai.core.worlds import create_task
+        from parlai.core.params import ParlaiParser
+
+        basename = basename = f"{self.task}_{datatype}".replace(":", "_")
```
Comment (EricMichaelSmith) [inline parlai/utils/testing.py:427]: Hmm looks like you have `basename =` twice

File: tests/datatests/test_new_tasks.py [hunk 4]
```diff
@@ -1,84 +0,0 @@
-#!/usr/bin/env python3
```
Comment (EricMichaelSmith) [inline tests/datatests/test_new_tasks.py:1]: 🎉

**Review by EricMichaelSmith — APPROVED** [review]
Seems reasonable - minor comments

**Pull request merged**